server/src/test/java/org/elasticsearch/cluster/service/MasterServiceTests.java

```diff
@@ -464,23 +464,39 @@ public void onFailure(Exception e) {
     }
 
     public void testClusterStateBatchedUpdates() throws BrokenBarrierException, InterruptedException {
```

Comment:
One important thing that this test seems not to check is that batches submitted with the same executor are executed together. It's a bit tricky to test because you have to block the master service while you're submitting the batches to make sure that they all arrive "at once". Still, I think this is something worth addressing either randomly in this test or else in a separate test.

Reply:
On reflection I think it doesn't make sense to try and extend this test to cover this extra invariant too. Let's leave it as it is and add a new test in a follow-up.

Reply:
I believe that is asserted here: https://github.com/idegtiarenko/elasticsearch/blob/665c374b90098f4dd34bcf238de21342a2b4c47a/server/src/test/java/org/elasticsearch/cluster/service/MasterServiceTests.java#L538-L545 (unless I miss something)

Reply:
This asserts that every set of tasks that was _submitted_ together is executed as an atomic unit. That's certainly important but in practice we only submit multiple tasks in `CandidateJoinAccumulator` when completing an election. We also want to know that tasks submitted with _multiple_ calls to `submitStateUpdateTask` are executed as a batch.